allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -270,36 +277,52 @@ def take_step(self,
+            # (batch_size, input_sequence_length)
+            question_attention_weights = self._input_attention(decoder_hidden,
+                                                               state.encoder_outputs,
+                                                               state.encoder_output_mask)
+            # (batch_size, encoder_output_dim)
+            attended_question = util.weighted_sum(state.encoder_outputs, question_attention_weights)
+
+        # (batch_size, decoder_input_dim)
+        decoder_input = self._input_projection_layer(torch.cat([attended_question, embedded_input], -1))
+
         decoder_hidden, decoder_context = self._decoder_cell(decoder_input,
                                                              (decoder_hidden, decoder_context))
 
+        # (batch_size, question_length)
+        question_attention_weights = self._input_attention(decoder_hidden,
+                                                           state.encoder_outputs,
+                                                           state.encoder_output_mask)
+        # (batch_size, encoder_output_dim)
+        attended_question = util.weighted_sum(state.encoder_outputs, question_attention_weights)
+
         # (batch_size, num_actions)
-        logits = self._output_projection_layer(decoder_hidden)
+        logits = self._output_projection_layer(torch.cat([decoder_hidden, attended_question], -1))
         output_mask = state.get_output_mask()
         log_probs = util.masked_log_softmax(logits, output_mask)
-        sorted_log_probs = []
-        # TODO(mattg): think about how to make batching work here; batch_size is currently required
-        # to be 1, so we enumerate `log_probs[0]`.
-        for i, log_prob in enumerate(log_probs[0]):
-            # We need to keep the gradient history, but also have a float for easy comparisons.
-            # That's why we have three things here: (float_log_prob, variable_log_prob, action_id).
-            sorted_log_probs.append((log_prob.data[0], log_prob, i))
-        sorted_log_probs.sort(key=lambda x: -x[0])
+        new_hidden_state = (decoder_hidden, decoder_context, attended_question)
+        return self._yield_new_states(state, log_probs, new_hidden_state, allowed_actions)
 
+    def _yield_new_states(self,  # pylint: disable=no-self-use
+                          state: WikiTablesDecoderState,
+                          log_probs: Variable,
+                          new_hidden_state: Tuple[Variable, Variable],
+                          allowed_actions: Set[int]) -> Generator[WikiTablesDecoderState, None, None]:
+        sorted_log_probs, sorted_actions = log_probs.sort(dim=-1, descending=True)
+        sorted_actions = sorted_actions.data.cpu().numpy().tolist()
         valid_actions = state.get_valid_actions()
-        for _, log_prob, action in sorted_log_probs:
+        for i in range(sorted_log_probs.size(1)):
+            action = sorted_actions[0][i]
```

Comment:
This is `sorted_actions[0][i]` just because we have a batch size of 1, right? Could use a comment.

Reply:
Done.